gtsam/sfm/PowerMethod.h

```diff
@@ -0,0 +1,367 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010-2019, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file   PowerMethod.h
+ * @date   Sept 2020
+ * @author Jing Wu
+ * @brief  accelerated power method for fast eigenvalue and eigenvector computation
+ */
+
+#pragma once
+
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/Vector.h>
+#include <gtsam/dllexport.h>
+
+#include <Eigen/Core>
+#include <Eigen/Sparse>
+#include <map>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include <algorithm>
+#include <cmath>
+#include <chrono>
+#include <random>
+
+namespace gtsam {
+
+using Sparse = Eigen::SparseMatrix<double>;
+
+/* ************************************************************************* */
+/// MINIMUM EIGENVALUE COMPUTATIONS
+
+struct PowerFunctor {
+  /**
+   * \brief Computer i-th Eigenpair with power method
+   *
+   * References :
+   * 1) Rosen, D. and Carlone, L., 2017, September. Computational
+   * enhancements for certifiably correct SLAM. In Proceedings of the
+   * International Conference on Intelligent Robots and Systems.
+   * 2) Yulun Tian and Kasra Khosoussi and David M. Rosen and Jonathan P. How,
+   * 2020, Aug, Distributed Certifiably Correct Pose-Graph Optimization, Arxiv
+   * 3) C. de Sa, B. He, I. Mitliagkas, C. Ré, and P. Xu, “Accelerated
+   * stochastic power iteration,” in Proc. Mach. Learn. Res., no. 84, 2018, pp. 58–67
+   *
+   * It performs the following iteration: \f$ x_{k+1} = A * x_k + \beta *
+   * x_{k-1} \f$ where A is the certificate matrix, x is the ritz vector
+   *
+   */
+
+  // Const reference to an externally-held matrix whose minimum-eigenvalue we
+  // want to compute
+  const Sparse &A_;
+
+  const Matrix &S_;
+
+  const int m_n_; // dimension of Matrix A
+  const int m_nev_; // number of eigenvalues required
```

Comment:
Similarly, why is this needed?  (The power method only ever returns a **single** Ritz pair ... )

Reply:
Yes, the power method only returns a single Ritz pair. m_nev_ is used when we want to sort all ritz vectors.

Reply:
But the power method only computes a **single** Ritz vector (the one that gets returned when the function terminates).  So what are these other not-Ritz-vectors that we're keeping around, and why would be interesting in "sorting" them?